### ISSUE TITLE
Please update rustdb-rocksdb version including formats

### ISSUE THREAD

**Issue by ytrezq** [openethereum/openethereum#329]
OpenEthereum is still downloading from https://github.com/paritytech/rust-rocksdb.

Newer versions bring performance improvements like a better handling of numa.

Administration tools offer from newer rustdb-rocksdb supports to replace damaged sst files (see #264) but they update the Rocksdb version so OpenEthereum can’t use it.

**Comment by yorickdowne** [comment]
Probably see https://github.com/openethereum/openethereum/pull/270 ? That upstream PR has been accepted into `master` and now needs to make it into a release.

**Comment by ytrezq** [comment]
@yorickdowne I don t know. Did it update the underlying rocksdb c++ libraries and database format?

**Comment by rakita** [comment]
It does update underlying rocksdb, but this is the point of the pull request, it does not update any database layout and it is backward and even forward compatible on the current database. Why it is still not merged, It was dependent on one bigger pull request that upgrades ethereum-types that is merged last week. We need to detangle merge conflict, do some testing to see if harddisk memory is stable, and do the merge.

**Comment by rakita** [comment]
I will close it and marge it as a duplicate of https://github.com/openethereum/openethereum/issues/89

**Comment by ytrezq** [comment]
@rakita then not a duplicate. Incompatible format change introduced by newer version of rocksdb are driven toward higher peformance like higher cpu cache hit in the case of bloom filters.

**Comment by rakita** [comment]
I am not sure if I follow, upgrading `kvdb-*` is basically upgrading `rocksdb`, and as mdben reported here https://github.com/openethereum/openethereum/issues/89#issuecomment-778026072 performance goes up. btw bloom filters collum in db is removed last year.

**Comment by ytrezq** [comment]
@rakita yes but with the newer formats. It could even go faster. And as a bonus, this open compatibility with official maintenance tools for manual database recovery (like adding missing damaged keys).

**Comment by rakita** [comment]
Yea, I see what you mean: https://rocksdb.org/blog/2019/03/08/format-version-4.html
will need to check

**Comment by ytrezq** [comment]
> Yea, I see what you mean: https://rocksdb.org/blog/2019/03/08/format-version-4.html
> will need to check

No. I was talking about version 5 not 4.

**Comment by ytrezq** [comment]
> It does update underlying rocksdb, but this is the point of the pull request, it does not update any database layout and it is backward and even forward compatible on the current database. Why it is still not merged, It was dependent on one bigger pull request that upgrades ethereum-types that is merged last week. We need to detangle merge conflict, do some testing to see if harddisk memory is stable, and do the merge.

@rakita no. I just tried the https://github.com/openethereum/openethereum/pull/270 pull request and indeed older versions or current master branch can’t use the database anymore as OpenEthereum stops with:

    /usr/share/rust/.cargo/registry/src/github.com-1ecc6299db9ec823/parity-rocksdb-sys-0.5.6/rocksdb/db/db_impl_compaction_flush.cc:1953] Compaction error: Corruption: Unknown Footer version. Maybe this file was created with newer version of RocksDB?

So it’s not backward compatible.

**Comment by rakita** [comment]
@ytrezq  okay it seems it is not forward compatible from `newer`->`older` but it should be backward compatible `older`->`newer`. You probably can upgrade from `master` to #270 but not other way around. Will need to check those cases.

**Comment by rakita** [comment]
yes version 5 is probably what we want: https://github.com/facebook/rocksdb/blob/3bfd3ed2f393c8d659822fba0860b9dde22efb0b/include/rocksdb/table.h#L368-L394

**Comment by ytrezq** [comment]
> @ytrezq okay it seems it is not forward compatible from `newer`->`older` but it should be backward compatible `older`->`newer`. You probably can upgrade from `master` to #270 but not other way around. Will need to check those cases.

Isn’t rather `forward compatible` what you were really meaning?

**Comment by rakita** [comment]
> > @ytrezq okay it seems it is not forward compatible from `newer`->`older` but it should be backward compatible `older`->`newer`. You probably can upgrade from `master` to #270 but not other way around. Will need to check those cases.
> 
> Isn’t rather `forward compatible` what you were really meaning?

It is little bit not intuitive but backward compatible means you can use `new software` to run `old db`, while forward compatible means that you can run `new db`  on `old software`. here is example picture: https://tech.amikelive.com/wp-content/uploads/2018/12/cuda_driver_backward_compatibility_rev-768x388.png
